Update privacy toggle to be per site

subsymbolic opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: https://app.asana.com/0/72649045549333/559623300949085

**Steps to test this PR**:

**TEST 1: Test new toggle text and behavior**
1. Navigate to a page e.g cnn.com
1. Open the privacy dashboard and ensure that the toggle text is "Site Privacy Protection"
1. Switch privacy protection off and press back
1. Ensure page reloads
1. Ensure page is now in whitelist
1. Ensure that mp_wla pixel is fired
1. Navigate to another page e.g facebook.com
1. Check that privacy toggle is on
1. Navigate back cnn.com and ensure that the privacy/site protection toggle is back off
1. Switch the toggle back on
1. Ensure that  mp_wlr pixel is fired
1. Ensure page is no longer in whitelist

**TEST 2: Test new buttons**
1. Navigate to a page and open the privacy Dashboard
1. Tap the _Manage Whitelist_ button and ensure whitelist page is launched
1. Ensure the "mp_mw" pixel fires
1. Go back
1. Tap the _Report Broken Site_ button and ensure that the broken site button in launched
1. Ensure the "mp_rb" pixel fires

**TEST 3: Test new pixels browsing pixels**
1. Browse to a site
1. From the browsing menu, add it to the whitelist and ensure that "mb_wla" fires
1. From the browsing menu, remove it from the whitelist and ensure that "mb_wlr" fires

**TEST 4: Test new whitelist settings pixel**
1. Open _Settings_
1. Tap _Privacy Protection Whitelist_
1. Ensure "ms_mw" fires

**TEST 5: Test whitelist page**
1. Open _Settings_
1. Tap _Privacy Protection Whitelist_
1. Add, edit and delete entries and ensure these behave as expected
1. Go back to settings and switch theme
1. Now repeat step 3 to ensure UI still works as expected (note dialogs are universally white as per bookmarks UI)
1. Repeat this test on API 21`

**TEST 6: Test database migration**
1. Cleanly install the app from develop and run the app to ensure the db is initialized
1. No install this branch over the top to ensure
1. Run the app and ensure the migration succeeds without anything breaking

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

subsymbolic commented on app/src/main/res/layout/view_whitelist_entry.xml:
```diff
@@ -0,0 +1,65 @@
+        android:paddingStart="14dp"
+        android:paddingEnd="14dp"
+        android:scaleType="center"
+        android:src="@drawable/ic_overflow_24dp"
```
Ahh good catch, that's new, must have come in from a merge. That's fixed and I'll see if I can remove the new need to double define the color like this during LHF.

marcosholgado replied:
Haven't tried this and just commenting because I saw this. What about using `ImageViewCompat` and `app:srcCompat="@drawable/ic_overflow_24dp"` instead of using tint?

cmonfortep replied:
@marcosholgado I've just tried what you proposed and it works like a charm. At least we know there are two different solutions for this 👍

subsymbolic commented:
> Overflow icon color in whitelist page is rendered with pink color

Thanks! This was introduced by another PR and has subsequently been fixed globally by https://github.com/duckduckgo/Android/pull/814.

> Testing with destroy activities enabled: Having a whitelisted site, every time I visit "whitelist page" or "report broken site" or "trackers found"...mp_wla is sent without any interaction

Good catch! Fixed.

> Every time I visit privacy dashboard from a whitelisted site, I see the toggle moving from enabled to disabled

The visible ui update as the user enters the screen is preexisting behavior. We have had no feedback about and as we expect to either completely redesign or even delete this screen we have no plans to fix it.

> If I go to privacy dashboard and I switch the toggle twice (interacting but keeping the original state), and going back to the browser, page will reload. is that expected?

Yes. It's it is an easier implementation this way and I don't think users will find it odd that the page reloads after they have interacted with the toggle. I like your idea of being smarter and only reloading when the value has truly changed (on an odd number of toggles) however I don't think it's worth investing in that level of polish when we are soon deprecating this screen. If we receive feedback that this is unacceptable, we can reconsider. Let me know if you feel strongly otherwise.

> If I visit a website -> then I whitelist that site from settings -> I go back to the browser -> website doesn't reload. Is that something we expect? or should we detect the change and reload the site?

Yes this is expected. If the user is in the context of a site and they toggle protection either through the dash toggle or the whitelist menu button then we reload. If the go to whitelists screen via settings however and start adding and removing entries, the user is now outside the context of the current page. We could make this fancier but it also adds complexity. Let me know if you feel strongly about this though.

> When switching the site privacy protection toggle, should we change somehow the privacy grade to the user

Again this is preexisting behavior and as we expect to be deprecate this screen this isn't something we plan to change right now.

> Noticed we removed the trackers statistics from the privacy dashboard

Oh no that shouldn't happen, thanks for catching it. Fixed! Beware though there is another outstanding issue in develop in this area due to a change from percentage to ratio for prevalence in the backend API. To see any results you'll need to update `Entiry.kt` to have `MAJOR_NETWORK_PREVALENCE = 0.7`

cmonfortep commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:
```diff
@@ -497,6 +508,8 @@ class BrowserTabViewModel(
             canSharePage = true,
             showPrivacyGrade = true,
             canReportSite = true,
+            canWhitelist = canWhitelist,
```
I wanted to check what `canWhitelist` means vs `isWhitelisted`. Now I see that in Home page that value is `false`. But if we visit any navigate to any website and we press back (navigating to home page again), the value remains as `true` and the option enabled. This is minor due to the check `val domain = site?.domain ?: return`.

subsymbolic replied:
`canWhitelist` refers to whether the whitelisting feature is available similar to `canAddBookmarks` and `canShare`. `isWhitelisted` refers to if a site is actually whitelisted or not.

The issue you see was introduced globally at some recent point and will be fixed by https://app.asana.com/0/search/1175146161566013/1170812542831705. The menu state used to be reset and it no longer is which causes these quirks. But I can see that this is really icky and even more so when combined with this PR, so I'm going to pickup that issue as a priority. Thanks for flagging it!

cmonfortep replied:
Thanks for the asana reference, I was not aware of this one.

CDRussell commented on app/src/main/java/com/duckduckgo/app/privacy/ui/WhitelistViewModel.kt:
```diff
@@ -0,0 +1,109 @@
+            command.value = ShowWhitelistFormatError
+            return
+        }
+        GlobalScope.launch(dispatchers.io()) {
```
_I don't have the full context of the code which calls this function, so bear that in mind when considering my answer here._

Generally, the problem with `GlobalScope.launch` here, or any launching from any scope at all, is the risk of creating a race condition, whereby `onEntryAdded()` has finished executing but the launched coroutine is still executing. The risk is that we base logic on `onEntryAdded()` finishing and assume the entry has in fact been added, when in reality it might have been added already, might still to be added or be in the middle of being added; we have no way of knowing from the calling code.

if `onEntryAdded` were a `suspend` function, we'd have more control over this but I don't know how that would affect other code and if we need that control. we've been burnt a few times because of similar problems, so it would certainly be good to start tightening these things up instead of introducing more of them. But again, i'm not familiar with the larger changes introduced here. Happy to dig in more if required, or discuss offline if folks wanted to sync up about it.

cmonfortep replied:
Thanks @CDRussell, that's a very nice explanation for the things to take into consideration when using `launch`.

I was not proposing to replace `launch` but to change the `coroutineScope`. Is there any recommendation or things to validate before using `globalScope` instead of `viewmodelScope`?

subsymbolic replied:
We discussed this over zoom and concluded that:
* keeping GlobalScope is ok as we want the operation to live on past the 
* the db methods should be suspend functions for clarity and to avoid people using them without considering async nature as this method grows - this has been updated

marcosholgado commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:
```diff
@@ -507,9 +520,27 @@ class BrowserTabViewModel(
+    private fun updateLoadingStatePrivacy() {
+        viewModelScope.launch(dispatchers.io()) {
+            site?.domain?.let {
+                loadingViewState.postValue(currentLoadingViewState().copy(privacyOn = !userWhitelistDao.contains(it)))
```
Question about this. Are we only disabling the trackers animation/dax dialogs if the user has a specific website in the whitelist? What happens if they disable privacy protection overall or is that not possible anymore?

subsymbolic replied:
@marcosholgado there is no longer a global switch.

subsymbolic commented on app/src/androidTest/java/com/duckduckgo/app/httpsupgrade/HttpsUpgraderTest.kt:
```diff
@@ -68,6 +70,13 @@ class HttpsUpgraderTest {
+    @Test
+    fun whenHttpDomainIsUserWhitelistedThenShouldNotUpgrade() {
+        whenever(mockUserWhitelistDao.contains("www.local.url")).thenReturn(true)
+        bloomFilter.add("www.local.url")
```
No, we need to add the url to the bloom filter so it is in our upgrade list and then test that it's not upgraded despite this because of the user whitelist. The test should fail before the new user whitelist logic is added (red) and then pass after (green). If we remove adding it to the bloom filter then the test passes even before the new logic is added (or if we comment it out) meaning that it does nothing to test the new logic.

I like your pixel test suggestion. Added!

cmonfortep replied:
aha! I really enjoyed this explanation, you got me thinking about it for a while. I really liked the green/red reference 👍

subsymbolic commented on app/src/main/java/com/duckduckgo/app/privacy/ui/ScorecardViewModel.kt:
```diff
@@ -67,28 +76,35 @@ class ScorecardViewModel(private val settingsStore: PrivacySettingsStore) : View
         )
     }
 
+    @WorkerThread
```
Ah good point! Thanks :)

subsymbolic commented on app/src/main/java/com/duckduckgo/app/privacy/ui/WhitelistActivity.kt:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    class WhitelistViewHolder(
+        val root: View,
```
I could get rid of it but then I'd need to attach a more generic click listener and I think this is a little better as it eliminates some ugliness in that area. Resolving, since this is a nit, but happy to discuss it more if you have questions.

subsymbolic commented on app/src/main/java/com/duckduckgo/app/global/UriString.kt:
```diff
@@ -42,17 +43,16 @@ class UriString {
 
-            if (webUrlRegex.containsMatchIn(host)) return true
-            return false
+        fun isValidDomain(domain: String): Boolean {
```
I didn't as it is a convenience method to an android API, if it contained our own logic I definitely would. Let me know if you feel strongly that they are needed however.

cmonfortep replied:
Yes, my intention was kind of document the expected behavior for the regex used in there. This was already a pretty nice tidy up and I don't feel strongly about including the test. We are good 👍

subsymbolic commented on app/src/androidTest/java/com/duckduckgo/app/privacy/db/UserWhitelistDaoTest.kt:
```diff
@@ -0,0 +1,90 @@
+
+    @Test
+    fun whenInitializedThenListIsEmpty() {
+        assertTrue(dao.all().blockingObserve()!!.isEmpty())
```
They are both established patterns in our codebase and I tend to use whichever is most convenient for my current test suite. In this case, it was easier to observe the value directly but you make a good point about the observers and whether we need to reevaluate the pattern generally. I have created an Asana task https://app.asana.com/0/414730916066338/1175169310823907.

subsymbolic commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:
```diff
@@ -713,13 +744,33 @@ class BrowserTabViewModel(
+
+    fun onWhitelistSelected() {
+        val domain = site?.domain ?: return
+        GlobalScope.launch(dispatchers.io()) {
```
Update: We discussed this over zoom and concluded that GlobalScope makes sense if we want this operation to continue beyond the ViewModel scope.

subsymbolic commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:
```diff
@@ -713,13 +744,33 @@ class BrowserTabViewModel(
+    private fun addToWhitelist(domain: String) {
+        pixel.fire(PixelName.BROWSER_MENU_WHITELIST_ADD)
+        userWhitelistDao.insert(domain)
+        browserViewState.postValue(currentBrowserViewState().copy(isWhitelisted = true))
```
I like that, thanks!

subsymbolic commented on app/src/main/java/com/duckduckgo/app/privacy/ui/ScorecardViewModel.kt:
```diff
@@ -67,28 +76,35 @@ class ScorecardViewModel(private val settingsStore: PrivacySettingsStore) : View
-            showIsMemberOfMajorNetwork = site.entity?.isMajor ?: false,
-            showEnhancedGrade = grade != improvedGrade
-        )
+        viewModelScope.launch(dispatchers.io()) {
```
I agree - this is a good suggestion and a good pattern for us to employ. I have updated the code.

subsymbolic commented on app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:
```diff
@@ -507,9 +520,27 @@ class BrowserTabViewModel(
+    private fun updateLoadingStatePrivacy() {
+        viewModelScope.launch(dispatchers.io()) {
+            site?.domain?.let {
+                loadingViewState.postValue(currentLoadingViewState().copy(privacyOn = !userWhitelistDao.contains(it)))
```
I like it, thanks. Updated.

subsymbolic commented on app/src/main/java/com/duckduckgo/app/privacy/ui/PrivacyDashboardViewModel.kt:
```diff
@@ -46,26 +54,27 @@ class PrivacyDashboardViewModel(
         val allTrackersBlocked: Boolean,
         val practices: PrivacyPractices.Summary,
-        val toggleEnabled: Boolean,
+        val toggleEnabled: Boolean?,
```
Null means the state is still unknown (this happens briefly during when initializing before our db query has returned) and we render the toggle as on by default. It's a little ugly but the null is helpful as it lets us know if the state is unknown, true or false. If you feel strongly I can maintain that bit of state as a separate private ViewModel (not ViewState) variable. What do you think?

subsymbolic commented on app/src/main/java/com/duckduckgo/app/privacy/ui/WhitelistViewModel.kt:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    private suspend fun addEntryToDatabase(entry: UserWhitelistedDomain) {
+        withContext(dispatchers.io()) { dao.insert(entry) }
```
We don't really need to repeat the io dispatcher here as the launch already specifies an io dispatcher - however this method was added based on PR feedback around future safety and as it is possible that a future launch could be on the main thread, I have been explicit.

cmonfortep replied:
`withContext` internally checks if we are already in the same dispatcher, so there's no problem repeating here if that gives us more confidence, which I also agree is a good idea. Nice comment.

subsymbolic commented:
Thanks for the awesome review 🙏
